Author: Kurt Roeckx <kurt@roeckx.be>
Description: Reconnect after an EINVAL to make it work with dynamic IP addresses (Closes: #392250).

diff -Naurp openntpd.orig/client.c openntpd/client.c
--- openntpd.orig/client.c	2006-05-14 05:29:21.000000000 +0000
+++ openntpd/client.c	2008-01-15 17:08:58.000000000 +0000
@@ -116,6 +116,7 @@ int
 client_query(struct ntp_peer *p)
 {
 	int	tos = IPTOS_LOWDELAY;
+	int	result;
 
 	if (p->addr == NULL && client_nextaddr(p) == -1) {
 		set_next(p, error_interval());
@@ -163,9 +164,17 @@ client_query(struct ntp_peer *p)
 	p->query->msg.xmttime.fractionl = arc4random();
 	p->query->xmttime = gettime();
 
-	if (ntp_sendmsg(p->query->fd, NULL, &p->query->msg,
-	    NTP_MSGSIZE_NOAUTH, 0) == -1) {
+	if ((result = ntp_sendmsg(p->query->fd, NULL, &p->query->msg,
+	    NTP_MSGSIZE_NOAUTH, 0)) < 0) {
 		set_next(p, INTERVAL_QUERY_PATHETIC);
+		if (result == -2) {
+			/*
+			 * got EINVAL in sendto(), probably the local socket
+			 * address got invalidated -> force re-connect()
+			 */
+			close(p->query->fd);
+			p->query->fd = -1;
+		}
 		return (-1);
 	}
 
diff -Naurp openntpd.orig/ntp_msg.c openntpd/ntp_msg.c
--- openntpd.orig/ntp_msg.c	2006-05-14 05:29:21.000000000 +0000
+++ openntpd/ntp_msg.c	2008-01-15 17:08:58.000000000 +0000
@@ -98,6 +98,8 @@ ntp_sendmsg(int fd, struct sockaddr *sa,
 			return (-1);
 		}
 		log_warn("sendto");
+		if (errno == EINVAL)
+			return (-2);
 		return (-1);
 	}
 
